scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -42,37 +49,74 @@ class KRouter[A](val config: Config[A], val network: KNetwork[A])(
       )
   }
 
-  // Handle findNodes requests...
-  val nodes: ConnectableObservable[(FindNodes[A], Nodes[A] => Task[Unit])] = network.findNodes
-  nodes.foreach {
-    case (findNodesRequest, responseHandler) =>
-      val FindNodes(uuid, nodeRecord, targetNodeId) = findNodesRequest
-
+  network.kRequests.foreach {
+    case (FindNodes(uuid, nodeRecord, targetNodeId), responseHandler) =>
       debug(
         s"Received request FindNodes(${nodeRecord.id.toHex}, $nodeRecord, ${targetNodeId.toHex})"
       )
       add(nodeRecord)
-      val result = embellish(kBuckets.closestNodes(targetNodeId, config.k))
-      responseHandler(Nodes(uuid, config.nodeRecord, result)).runToFuture
-        .onComplete {
-          case Failure(t) =>
-            log.info(
-              s"Nodes response $uuid to ${nodeRecord.routingAddress} failed with exception: $t"
-            )
-          case _ =>
-        }
+
+      val result = Nodes(uuid, config.nodeRecord, embellish(kBuckets.closestNodes(targetNodeId, config.k)))
+
+      sendResponse(result, responseHandler, nodeRecord.routingAddress)
+    case (Ping(uuid, nodeRecord), responseHandler) =>
+      debug(
+        s"Received request Ping(${nodeRecord.id.toHex}, $nodeRecord)"
+      )
+      add(nodeRecord)
+      sendResponse(Pong(uuid, config.nodeRecord), responseHandler, nodeRecord.routingAddress)
+  }
+
+  private def sendResponse(response: KResponse[A], responseHandler: Option[KResponse[A]] => Task[Unit], to: A): Unit = {
+    responseHandler(Some(response)).runToFuture
+      .onComplete {
+        case Failure(t) =>
+          log.info(
+            s"Pong response ${response.requestId} to $to failed with exception: $t"
+          )
+        case _ =>
+      }
   }
-  nodes.connect()
 
   def get(key: BitVector): Future[NodeRecord[A]] = {
     debug(s"get(${key.toHex})")
     getLocally(key).recoverWith { case _ => getRemotely(key) }.recoverWith { case t => giveUp(key, t) }
   }
 
-  def add(nodeRecord: NodeRecord[A]): Unit = {
-    kBuckets.add(nodeRecord.id)
-    nodeRecords.put(nodeRecord.id, nodeRecord)
-    debug(s"Added record (${nodeRecord.id.toHex}, $nodeRecord) to the routing table.")
+  def add(nodeRecord: NodeRecord[A]): Unit = this.synchronized {
+
+    val iBucket = kBuckets.iBucket(nodeRecord.id)
+    val bucket = kBuckets.bucket(iBucket)
+    debug(s"Handling potential addition of candidate (${nodeRecord.id.toHex}, $nodeRecord) to ibucket $iBucket.")
+    debug(s"iBucket($iBucket) = $bucket")
+    if (bucket.size < config.k) {
+      bucket.add(nodeRecord.id)
+      nodeRecords.put(nodeRecord.id, nodeRecord)
+    } else {
+      // ping the bucket's least recently seen node (i.e. the one at the head) to see what to do
+      val recordToPing = nodeRecords(bucket.head)
+      network.ping(recordToPing, Ping(uuidSource(), config.nodeRecord)).runToFuture.onComplete {
```

Comment:
1. Note: the part running inside of `onComplete` is no longer synchronized. This is important, as `TimeSet` is not thread safe.
2. Even if you add synchronization inside, then two `add` operations might choose the same `bucket.head`, leading to incorrect results, such as attempting to remove it twice (which is not a problem) and adding two new  elements to the bucket (which is a problem, as bucket is growing.

The solution here might be using `Await` here, but that would mean `add` function would synchronously wait for ping result, which might take long time, blocking the operation that called `add`. Another would be keeping global queue / per-bucket queues of elements that were requested to be added and processing them in an actor-like manner.

I'll take a better look tomorrow, analysing interplay between different parts of code.

Reply:
Thinking about this a little more making it blocking would slow down processing requests. I believe we need a queue of elements to be added, processed as a stream (in an actor-like way). We can schedule a call to talk this over if you want.

Reply:
I've (1) updated TimeSet and KBuckets to make them persistent data structures. The router, which remains mutable, then has an AtomicReference to the latest KBucket state.  (2) rechecked the bucket size condition inside the `onComplete`. Unless I've overlooked something, this should then give threadsafe access to the buckets and consistent reads of the bucket state.